tests/src/test/java/alluxio/master/journal/ufs/UfsJournalLogWriterTest.java

```diff
@@ -178,6 +198,119 @@ public void writeJournalEntryRotate() throws Exception {
+      Assert.assertNull(writer.getJournalOutputStream());
+      Assert.assertEquals(10, writer.getNumberOfJournalEntriesToFlush());
+    }
+    // Perform another write. UfsJournalLogWriter will perform recover first.
+    writer.write(newEntry(nextSN));
+    nextSN++;
+    writer.close();
+
+    try (JournalReader reader = new UfsJournalReader(mJournal, startSN, true)) {
+      Journal.JournalEntry entry;
+      long expectedSeq = startSN;
+      while ((entry = reader.read()) != null) {
+        Assert.assertEquals(expectedSeq, entry.getSequenceNumber());
+        expectedSeq++;
+      }
+      Assert.assertEquals(expectedSeq, nextSN);
+    }
+  }
+
+  /**
+   * Test the case in which there are missing journal entries between the last persisted entry
+   * and the first entry in {@code UfsJournalLogWriter#mEntriesToFlush}.
+   * {@code UfsJournalLogWriter} should be able to detect this issue.
+   */
+  @Test
+  public void missingJournalEntries() throws Exception {
+    Mockito.when(mUfs.supportsFlush()).thenReturn(true);
+    long dummySN = 0x10;
+    UfsJournalLogWriter writer = new UfsJournalLogWriter(mJournal, dummySN);
+    UfsJournalSnapshot snapshot;
+    UfsJournalFile journalFile;
+    long truncateSize = 0;
+    long firstCorruptedEntrySeq = 0;
+    for (int i = 0; i < 5; i++) {
+      writer.write(newEntry(dummySN));
+      dummySN++;
+      if (i == 3) {
+        firstCorruptedEntrySeq = dummySN;
+        snapshot = UfsJournalSnapshot.getSnapshot(mJournal);
+        journalFile = snapshot.getCurrentLog(mJournal);
+        File file = new File(journalFile.getLocation().toString());
+        truncateSize = file.length();
+      }
+    }
+    writer.flush();
+
+    writer.write(newEntry(dummySN));
+    long seqOfFirstEntryToFlush = dummySN;
+    dummySN++;
+    Assert.assertNotNull(writer.getJournalOutputStream());
+    DataOutputStream badOut = Mockito.mock(DataOutputStream.class);
+    Mockito.doThrow(new IOException("injected I/O error")).when(badOut)
+        .write(Mockito.any(byte[].class), Mockito.anyInt(), Mockito.anyInt());
+    UfsJournalLogWriter.JournalOutputStream journalOutputStream = writer.getJournalOutputStream();
+    Whitebox.setInternalState(journalOutputStream, "mOutputStream", badOut);
+    Assert.assertEquals(writer.getUnderlyingDataOutputStream(), badOut);
+    try {
+      writer.write(newEntry(dummySN));
+      dummySN++;
+      // Should not reach here
+      Assert.fail();
+    } catch (IOException e) {
+      Assert.assertTrue(e.getMessage().contains("injected I/O error"));
+      Assert.assertNotEquals(journalOutputStream, writer.getJournalOutputStream());
+      Assert.assertNull(writer.getJournalOutputStream());
+      Assert.assertEquals(1, writer.getNumberOfJournalEntriesToFlush());
+    }
+    snapshot = UfsJournalSnapshot.getSnapshot(mJournal);
+    journalFile = snapshot.getCurrentLog(mJournal);
+    File file = new File(journalFile.getLocation().toString());
```

Comment:
we already set these earliler

Reply:
Is it recommended to obtain the current snapshot whenever we want to access the journal? Snapshot may have changed between now and previous invocation. Correct this if I'm wrong. Thanks!

Reply:
+1 for not assuming the snapshot is the same. If we aren't re-using the value set on line 266, the `snapshot` on line 266 should be declared at the time of assignment on line 266, and this `snapshot` should be declared here.

Reply:
Done.